shared/js/dataDriver.js

```diff
@@ -0,0 +1,283 @@
+        endAt: `"chat!${this.makeId(endTime + 1)}"`,
+        // Always have a limit to get things sorted correctly.
+        limitToLast: limit > 0 ? limit : this.MAX_LIMIT
+      });
+      return this._request("GET", this._buildUrl(), query);
+    }
+
+    /**
+     * Update a record additionally storing the timestamp.
+     */
+    update(type, id, value) {
+      let key = `${type}!${id}`;
+      return this._request("PUT", this._buildUrl(key), {
+        "timestamp": {
+          ".sv": "timestamp"
+        },
+        value: value
+      });
+    }
+
+    /** ****************** **
+     *** Internal Helpers ***
+     ** ****************** **/
+
+    /**
+     * Convert an object with key/value pairs into a query string.
+     */
+    _buildQuery(object) {
+      return Object.keys(object).map(key => `${key}=${object[key]}`).join("&");
+    }
+
+    /**
+     * Get a REST endpoint for a resource defaulting to the current room.
+     */
+    _buildUrl(resource = "", roomToken = this._roomToken) {
+      return `${this.BASE_FIREBASE}/${roomToken}/${resource}.json`;
+    }
+
+    /**
+     * Connect to a specified room for streaming updates and future requests.
+     */
+    _connectToRoom(token) {
+      this._roomToken = token;
+      this._sse = new EventSource(this._buildUrl() + "?" + this._buildQuery({
+        limitToLast: 1,
+        orderBy: '"timestamp"'
+      }));
+      this._sse.addEventListener("open", this);
+      this._sse.addEventListener("put", this);
+
+      // Determine the clock skew before continuing initialization.
+      this.update("meta", "lastConnect").then(({ timestamp }) => {
+        this._clockSkew = timestamp - Date.now();
+      }).then(() => {
+        let now = this.getServerTime();
+        this.requestChat(now - 24 * 60 * 60 * 1000, now, 25).then(result =>
+          this._processRecords(result));
+      });
+    }
+
+    /**
+     * Handle a single record dispatching appropriate actions.
+     */
+    _processRecord(key, data) {
+      let [, type] = key.match(/^([^!]+)!(.+)$/);
+      switch (type) {
+        case "chat": {
+          let chat = data.value;
+          chat.receivedTimestamp = new Date(data.timestamp).toISOString();
+          this._dispatcher.dispatch(new actions.ReceivedTextChatMessage(chat));
+          break;
+        }
+      }
+    }
+
+    /**
+     * Process multiple records.
+     */
+    _processRecords(records) {
+      if (records !== null) {
+        Object.keys(records).map(key => this._processRecord(key, records[key]));
+      }
+    }
+
+    /**
+     * Make a HTTP/REST server request.
+     */
+    _request(method, url, data) {
+      let isGet = method === "GET";
+      let body = isGet ? undefined : data;
+      let query = isGet && data ? `?${data}` : "";
+      return new Promise((resolve, reject) => {
+        let request = new XMLHttpRequest();
+        request.open(method, url + query);
+        request.onload = () => {
```

Comment:
Added tests. Ran into some issues for the reject case as LoopMochaUtils changes reject to throw.